### PR TITLE
common.xml: add smart_battery_info message

### PR DESCRIPTION
**Pull request by hendjoshsr71**
Add smart_battery_info message.

One small documentation change on the MAVLink side here: https://github.com/mavlink/mavlink/pull/1551

### CODE REVIEW

**Comment by peterbarker** [comment]
@WickedShell

**Comment by tridge** [comment]
we'll let @WickedShell have a look at this

**Comment by tridge** [comment]
looks good to me

File: message_definitions/v1.0/common.xml
Comment (tridge) [inline message_definitions/v1.0/common.xml:4952]: if this is per-cell, then where do we say the number of cells? I think we should have a uint8_t num_cells

**Review by tridge — CHANGES_REQUESTED** [review]

File: message_definitions/v1.0/common.xml
Comment (hendjoshsr71) [reply message_definitions/v1.0/common.xml:4952]: Makes sense to add that. The other item I thought about is discharge_maximum_current?

Comment (auturgy) [reply message_definitions/v1.0/common.xml:4952]: Remember that this message doesn't replace     <message id="147" name="BATTERY_STATUS"> but supplements it, so I think number of cells should be pulled out of the cell voltages sent in that message.  Discharge_maximum_current is potentially problematic, as it varies with voltage, and smart batteries should be managing this anyway.  I agree it would be useful in some circumstances though.

Comment (hendjoshsr71) [reply message_definitions/v1.0/common.xml:4952]: Discharge_current_maximum doesn't vary with voltage as it is a manufacturer specification.  Perhaps you meant discharge_maximum_power? Only reason I bring it up is that this message and  message 147 seems to capture most of the important info. regarding a pack except the max. discharge current. And I saw the use of a few parameters to limit motor power based upon maximum discharge.

Comment (WickedShell) [reply message_definitions/v1.0/common.xml:4952]: Notably nothing in the message seems to tell you what maximum voltage is...

Comment (auturgy) [reply message_definitions/v1.0/common.xml:4952]: I think the intent was that max voltage is inferred from chemistry and number of cells, but valid point.
We’re starting to scope creep though.

Comment (hendjoshsr71) [reply message_definitions/v1.0/common.xml:4952]: I'm wondering if the charging_minimum_voltage was really meant to be charging_maximum_voltage? It is a little odd for an autopilot to care about that as usually is it only for chargers to prevent charging of a deeply discharged battery.

Comment (peterbarker) [reply message_definitions/v1.0/common.xml:4952]: @hendjoshsr71 on the min-voltage thing - hybrid vehicles are a thing...

**Comment by peterbarker** [comment]
Are we going to suggest adding that maximum-voltage thing upstream?

**Comment by hendjoshsr71** [comment]
Organizing common/important battery specs not in this message currently for discussion.
- `charging_maximum_voltage`
- `discharge_maximum_current`
- `discharge_burst_current`   (somewhat covered by previous)
- `manufacture_date` (could concatenate this into product name? From looking through specs and TI ICs `manufacturer_name` and `device_name` are a max. of 12 chars each )

**Comment by hendjoshsr71** [comment]
Added all of the items above to the message and created an upstream PR https://github.com/mavlink/mavlink/pull/1618

File: message_definitions/v1.0/common.xml
Comment (hendjoshsr71) [reply message_definitions/v1.0/common.xml:5099]: Added these items

**Comment by hamishwillee** [comment]
FYI, I reviewed in common - https://github.com/mavlink/mavlink/pull/1618#issuecomment-824547417 . My main concern is whether there are any known implementations that will be broken by this (yes it is marked WIP, but as you all know, we're trying to have common only for truly solid stuff. So if this has no implementations it really should be in development.xml.

**Comment by hendjoshsr71** [comment]
Rebased and changed the name of `num_cells` to `cells_in_series` for clarity 

@hamishwillee I have a PR in progress to use this message here https://github.com/ArduPilot/ardupilot/pull/16167.

**Comment by hamishwillee** [comment]
> Rebased and changed the name of num_cells to cells_in_series for clarity

@hendjoshsr71 Sounds like a good change. Note I am assuming that you're pushing these upstream into mavlink/mavlink since that is what get's signed off in the repo.

**Comment by hendjoshsr71** [comment]
Just pushed these changes to mavlink/mavlink. I will update the ArduPilot side with the manufacture data format that comes out from the Mavlink Dev call.

**Comment by hendjoshsr71** [comment]
Made these new fields extensions, removed WIP tags, and switched to `Maximum per-cell voltage when charged` using an uint16_t

**Comment by hamishwillee** [comment]
@peterbarker The upstream version has some minor differences to this, but I think we're pretty happy with it. Can you please confirm it also makes sense to you/"the ardupilot community" before I merge. Otherwise I'll wait for James to review in devcall next week.

**Comment by hendjoshsr71** [comment]
I haven't updated this version yet because I was waiting to get the upstream side settled first. :)

**Comment by hamishwillee** [comment]
@hendjoshsr71 That's fair enough except that what will ideally happen is that someone will pull your changes down from mavlink/mavlink rather than merge this. But we still want the message to make sense to everyone - so I'm hoping Peter can check the mavlink version.

**Comment by hamishwillee** [comment]
Well, it got merged, so you can either sync to this or wait for it to come from upstream.

**Review by tridge — APPROVED** [review]

**Pull request merged**
